src/cascadia/TerminalConnection/ConhostConnection.cpp

```diff
@@ -190,38 +193,75 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
     DWORD ConhostConnection::_OutputThread()
     {
         const size_t bufferSize = 4096;
-        BYTE buffer[bufferSize];
-        DWORD dwRead;
+        BYTE buffer[bufferSize]{ 0 };
+        BYTE utf8Partials[4]{ 0 }; // buffer for code units of a partial UTF-8 code point that have to be cached
+        DWORD dwRead{}; // length of a chunk read
+        DWORD dwPartialsLen{}; // number of cached UTF-8 code units
+        bool fSuccess{};
+        const BYTE bitmasks[]{ 0, 0b11000000, 0b11100000, 0b11110000 }; // for comparisons after the Lead Byte was found
+
+        // process the data of the output pipe in a loop
         while (true)
         {
-            dwRead = 0;
-            bool fSuccess = false;
+            // copy UTF-8 code units that were remaining from the previously read chunk (if any)
+            if (dwPartialsLen != 0)
+            {
+                std::move(utf8Partials, utf8Partials + dwPartialsLen, buffer);
+            }
+
+            // try to read data
+            fSuccess = !!ReadFile(_outPipe.get(), &buffer[dwPartialsLen], std::extent<decltype(buffer)>::value - dwPartialsLen, &dwRead, nullptr);
+
+            dwRead += dwPartialsLen;
+            dwPartialsLen = 0;
 
-            fSuccess = !!ReadFile(_outPipe.get(), buffer, bufferSize, &dwRead, nullptr);
-            if (!fSuccess)
+            if (dwRead == 0) // quit if no data has been read and no cached data was left over
+            {
+                return 0;
+            }
+            else if (!fSuccess)
             {
                 if (_closing.load())
                 {
                     // This is okay, break out to kill the thread
                     return 0;
                 }
-                else
-                {
-                    _disconnectHandlers();
-                    return (DWORD)-1;
-                }
+
+                _disconnectHandlers();
+                return (DWORD)-1;
             }
-            if (dwRead == 0)
+
+            const BYTE* const endPtr{ buffer + dwRead };
+            const BYTE* backIter{ endPtr - 1 };
+            // If the last byte in the buffer was a byte belonging to a UTF-8 multi-byte character
+            if ((*backIter & 0b10000000) == 0b10000000)
             {
-                continue;
+                // Check only up to 3 last bytes, if no Lead Byte was found then the byte before must be the Lead Byte and no partials are in the buffer
+                for (DWORD dwSequenceLen{ 1UL }, stop{ dwRead < 4UL ? dwRead : 4UL }; dwSequenceLen < stop; ++dwSequenceLen, --backIter)
+                {
+                    // If Lead Byte found
+                    if ((*backIter & 0b11000000) == 0b11000000)

```

Comment:
Can we get some names for these bit masks? I'd be surprised if they're not somewhere else in the codebase already.

Reply:
At least I'm not aware of any common names. [Wikipedia description for the meaning of these bits.](https://en.wikipedia.org/wiki/UTF-8#Description)
But I'll search the code base for these values. Maybe you already defined names that could be re-used here. (Perhaps in an `enum`?)

Reply:
https://github.com/microsoft/terminal/blob/9b92986b49bed8cc41fde4d6ef080921c41e6d9e/src/host/utf8ToWideCharParser.cpp#L15-L20
Different names for the same value. Doesn't include all values used here.  
<br>

https://github.com/microsoft/terminal/blob/9b92986b49bed8cc41fde4d6ef080921c41e6d9e/src/host/ut_host/Utf8ToWideCharParserTests.cpp#L296-L298
Better. But names are only in the comments.

Reply:
I'm personally okay without the names :smile:

Reply:
I can't agree more.  
Yesterday I fiddled with an `enum` and member names like `LeadByteTwoByteSequence`. Then I used these names in the array and noticed that they are ambiguous. E.g. we use value `0b11100000` as bitmask in the AND operation and compare the result with `0b11000000`. In the end each of these values would need two names. For example something like `MaskLeadByteTwoByteSequence` and `IsLeadByteThreeByteSequence` for value `0b11100000`. Since we can't use two names for an array element, we would need a second array to disambiguate the element names. I don't like it because it's a mess and it doesn't improve the readability imo. When I tried it yesterday I actually confused myself...  
In contrast, the binary literals are quite self-explanatory.